source/common/common/inline_map.h

```diff
@@ -0,0 +1,459 @@
+     * key/value pair in the inline map without hash searching.
+     */
+    uint64_t inlineId() const { return inline_id_; }
+
+  private:
+    friend class InlineMapDescriptor;
+
+    // This constructor should only be called by InlineMapDescriptor.
+    InlineKey(uint64_t inline_id) : inline_id_(inline_id) {}
+
+    uint64_t inline_id_{};
+
+#ifndef NDEBUG
+  public:
+    // This is the descriptor id that the inline key created by. This is used to validate the
+    // inline key validity.
+    uint64_t descriptorId() const { return descriptor_id_; }
+
+  private:
+    InlineKey(uint64_t inline_id, uint64_t descriptor_id)
+        : inline_id_(inline_id), descriptor_id_(descriptor_id) {}
+    uint64_t descriptor_id_{};
+#endif
+  };
+
+  using InlineKeysMap = absl::flat_hash_map<Key, uint64_t>;
+  using InlineKeys = std::vector<Key>;
+
+  InlineMapDescriptor() = default;
+
+  /**
+   * Add and create an custom inline key. May only be called before finalize().
+   */
+  template <class ArgK> InlineKey addInlineKey(const ArgK& key) {
+    RELEASE_ASSERT(!finalized_, "Cannot create new inline key after finalize()");
+
+    uint64_t inline_id = 0;
+
+    if (auto it = inline_keys_map_.find(key); it != inline_keys_map_.end()) {
+      // If the key is already added, return the existing key.
+      inline_id = it->second;
+    } else {
+      // If the key is not added yet, then create a new inline key.
+      inline_id = inline_keys_map_.size();
+      inline_keys_.push_back(Key(key));
+      inline_keys_map_[key] = inline_id;
+    }
+
+#ifndef NDEBUG
+    return {inline_id, descriptor_id_};
+#else
+    return {inline_id};
+#endif
+  }
+
+  /**
+   * Fetch the inline key for the given key. May only be called after finalize(). This should be
+   * used to get the inline key for the key added by addInlineKey(). This function could used to
+   * determine if the key is added as inline key or not at runtime or xDS config loading time and
+   * decide if the key should be used as inline key or normal key.
+   */
+  template <class ArgK> absl::optional<InlineKey> getInlineKey(const ArgK& key) const {
+    ASSERT(finalized_, "Cannot get inline handle before finalize()");
+
+    uint64_t inline_id = 0;
+
+    if (auto it = inline_keys_map_.find(key); it == inline_keys_map_.end()) {
+      return absl::nullopt;
+    } else {
+      inline_id = it->second;
+    }
+
+#ifndef NDEBUG
+    return InlineKey{inline_id, descriptor_id_};
+#else
+    return InlineKey{inline_id};
+#endif
+  }
+
+  /**
+   * Get the inline keys map that contains all inline keys and their index. May only be called after
+   * finalize().
+   */
+  const InlineKeysMap& inlineKeysMap() const {
+    ASSERT(finalized_, "Cannot fetch registration map before finalize()");
+    return inline_keys_map_;
+  }
+
+  /**
+   * Get the array that contains all added inline keys. May only be called after finalize().
+   */
+  const InlineKeys& inlineKeys() const {
```

Comment:
> handles() for consistency.

I think this should be `inlineKeys()` because the `Handle` is something to represent the inline key, `Handle`s are not the keys. But the `inlineKeys()` will return all added keys in original type.

> However I think we should not expose the vector object, and instead provide an iterate method on the descriptor. I don't think it will be used by anyone except the map's iterate method.

This is used mostly for debugging. For example, users could print all added inline keys in one line log by a simple `absl::StrJoin`.

Reply:
I think even in the release mode, we may still want to print the keys by `ENVOY_LOG(debug, ...)`. And I think it should be public because developers may want to iterate all these keys and print then in other modules?

Reply:
I'd like to keep the implementation-hiding strong. See https://www.hyrumslaw.com/.

I realize it's convenient to expose your representation but I'd prefer not to do it. You can expoe a debugString() if you want and I'm OK with not having that be debug-only if you prefer. You might want to expose that as an operator<<(std::ostream&) though and it can work only if the Key is also streamable.

Reply:
maybe a debugString is simple and enough.

Reply:
debugString is also fine; you can use ostringstream or similar to serialize it and require the key to support streaming.